pep-0722.rst

```diff
@@ -0,0 +1,271 @@
+   the format when installing their dependencies manually using a tool like pip.
+2. Use a tool that recognises and processes dependency blocks. This PEP does not
+   cover teaching users about such tools. It is assumed that if they are
+   popular, users will find out about them as with any other library or tool.
+
+Note that the core Python interpreter does *not* interpret dependency blocks.
+This may be a point of confusion for beginners, who try to run ``python
+some_script.py`` and do not understand why it fails. It is considered the
+responsibility of the person sharing the script to include clear instructions on
+how to run it.
+
+
+Reference Implementation
+========================
+
+This format is already supported `in pipx <https://github.com/pypa/pipx/pull/916>`__
+and in `pip-run <https://pypi.org/project/pip-run/>`__.
+
+
+Rejected Ideas
+==============
+
+Why not include other metadata?
+-------------------------------
+
+There is no obvious use case for other metadata, and if a project *does* need to
+specify anything more than some 3rd party dependencies, it has probably reached
+the point where it should be structured as a full-fledged project with a
+``pyproject.toml`` file.
+
+What about version?
+-------------------
+
+The one obvious exception is a script version number. The use cases for a
+version are, however, very different from those for dependencies, and it seems
+more reasonable to keep the two separate. There are already existing conventions
+for keeping a version number in a script (a ``__version__`` variable is a common
+approach) and these seem perfectly adequate.
+
+Why not make the dependencies visible at runtime?
+-------------------------------------------------
+
+This would typically involve storing the dependencies as a (runtime) list
+variable with a conventional name, such as::
+
+    __requires__ = [
+        "requests",
+        "click",
+    ]
+
+This has a number of problems compared to the proposed solution.
+
+1. The consumer has to parse arbitrary Python code, which almost certainly means
+   using the stdlib AST module, making it much harder for non-Python code to
+   read the data, as well as making Python code that does so significantly more
+   complex.
+2. Python syntax changes every version. While the requirement data only uses a
+   simple subset, the full file still needs to be parsed to *find* the
+   requirement data.
+3. This would reserve a specific global name (``__requires__``) in the above,
+   potentially clashing with user code.
+4. Users could assume that the value can be manipulated at runtime, and would
+   get unexpected results if they tried to do so.
+
+Furthermore, there is no known use case where being able to read the list of
+requirements at runtime is needed.
+
+It is worth noting, though, that the ``pip-run`` utility does implement (an
+extended form of) this approach. See `here <pip-run issue_>`_ for further
+discussion.
```

Comment:
Minor: best to avoid using just `here` for link texts.

Reply:
A suggested rephrase from Éric's point:

```suggestion
extended form of) this approach.
(`Further discussion <pip-run issue_>`_) is available on
their design of 'embedded requirements'.
```